src/incentives/AffinePass.sol

```diff
@@ -0,0 +1,182 @@
+        require(totalSupply() + amount <= MAX_SUPPLY, "Exceeds max supply");
+        require(reserveTokens + amount <= MAX_RESERVE_TOKENS, "Exceeds max reserve supply");
+        reserveTokens += amount;
+        for (uint256 i = 0; i < amount; i++) {
+            uint256 tokenId = _tokenIdCounter.current();
+            _tokenIdCounter.increment();
+            _safeMint(owner(), tokenId);
+        }
+    }
+
+    function hasRemainingSupply() public view returns (bool) {
+        uint256 currentSupply = totalSupply() - reserveTokens;
+        return currentSupply < MAXMINTABLESUPPLY;
+    }
+
+    function mintDrop(address[] memory recipients, uint256[] memory quantities) public onlyOwner {
+        require(recipients.length == quantities.length, "Recipients and quantities length mismatch");
+        for (uint256 i = 0; i < recipients.length; i++) {
+            for (uint256 j = 0; j < quantities[i]; j++) {
+                require(totalSupply() + quantities[i] <= MAX_SUPPLY, "Exceeds max supply");
+                uint256 tokenId = _tokenIdCounter.current();
+                _tokenIdCounter.increment();
+                _safeMint(recipients[i], tokenId);
+            }
+        }
+    }
+    
+    function mintWhitelist(bytes32[] memory proof) public payable {
+        require(
+                (whitelistSaleIsActive && isWhitelisted(_msgSender(), proof)),
+            "Sale paused or not whitelisted"
+        );
+        require(hasRemainingSupply(), "Exceeds max supply");
+        require(
+            _whitelistMinted[_msgSender()] + 1 <= MAX_WHITELIST_MINT,
+            "Exceeds max WL mint"
+        );
+
+        _whitelistMinted[_msgSender()] += 1;
+
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(_msgSender(), tokenId);
+    }
+
+    function mint() public payable {
+        require(saleIsActive, "Sale is not active");
+        require(hasRemainingSupply(), "Exceeds max supply");
+        require(
+            _minted[_msgSender()] + 1 <= MAX_PUBLIC_MINT,
+            "Exceeds max public mint"
+        );
+
+        _minted[_msgSender()] += 1;
+
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(_msgSender(), tokenId);
+
+    }
+
+    function withdraw() public onlyOwner {
+        uint256 balance = address(this).balance;
+        payable(owner()).transfer(balance);
+    }
+
+    function _baseURI() internal view virtual override returns (string memory) {
+        return baseURI;
+    }
+
+    function setBaseURI(string memory URI) public onlyOwner {
+        baseURI = URI;
+    }
+
+    // Overrides and Bridge
+
+    function supportsInterface(bytes4 interfaceId) public view override(ERC721, ERC721Enumerable) returns (bool) {
+        return super.supportsInterface(interfaceId);
+    }
+
+    function _beforeTokenTransfer(address from, address to, uint256 tokenId, uint256 batchSize)
```

Comment:
This `_beforeTokenTransfer` function in the version of OpenZeppelin we are [only uses three params](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/a7818b49a481e867230d4fa49d106bd304c5d95d/contracts/token/ERC721/ERC721.sol#L425). 

This doesn't build right now because of this issue.

Reply:
When I try to build and deploy with hardhat it forces me to use 4 params

Reply:
I can change it to 3 params in this pr just to get the build to run though